text/0000-unsafe-extern-blocks.md

```diff
@@ -0,0 +1,322 @@
+
+While we could, in a later edition, perhaps flip the default to make items safe to call, we could only do that if enough code has already been migrated.  But in the interim, we'd be asking users to accept the churn of migrating to this syntax without receiving any of the benefits.  That seems a bit like a cyclic dependency, so we've chosen not to do that.
+
+## Require all items to be marked as either `safe` or `unsafe`
+
+One could ask, why not require all items within an `unsafe extern` block to be marked as either `safe` or `unsafe` rather than making this optional?  Or alternatively, one could ask, why not *only* allow items to be marked as `unsafe` and *require* that all items be marked `unsafe`?
+
+As described in the last section, doing this would lead to a worse migration story for users, and so we chose not to do this.
+
+## Wait until we switch to `safe { .. }` blocks
+
+One could ask, why not wait to do this at all until we switch the language to use `safe { .. }` rather than `unsafe { .. }` blocks and then align this RFC with that?
+
+The problem with this is that there is no current plan to make such a switch.  Waiting to improve the language on a possibility that may or may not happen -- and in any case, will not happen soon -- is usually not a good plan.
+
+## Use `trusted` as the contextual keyword
+
+One could ask, why not use `trusted` rather than `safe` as the contextual keyword?  E.g.:
+
+```rust
+unsafe extern {
+    pub trusted fn sqrt(x: f64) -> f64; // Safe to call.
+    pub unsafe fn strlen(p: *const c_char) -> usize;
+}
+```
+
+The Rust language already has an accepted semantic for "safe" and "unsafe".  If we were to introduce a separated "trusted" concept, that would need to be part of a larger plan.  Such a plan does not yet exist in any concrete form, and it's not clear at this point whether any plan along these lines will succeed in gaining consensus.  Waiting to deliver value here on that possibility seems like a bad plan.
+
+If we later decide, e.g., to replace all uses of `unsafe { .. }` with `trusted { .. }`, large amounts of code would need to be changed in that migration.  Changing from `safe fn` to `trusted fn` as part of that, as this RFC would require, doesn't seem that it would make that migration markedly more painful.
+
+## Fire the `unsafe_code` lint for `extern` blocks also
+
+This RFC specifies that the `unsafe_code` lint will fire for `unsafe extern` but not for `extern` blocks.  One could ask, why not fire this for `extern` blocks also?
+
+The problem with doing this is that it may be very noisy on existing editions.  We're careful when expanding the meaning of existing lints to not create too much noise for existing code on existing editions, and doing this, at least immediately, may run afoul of this.
```

Comment:
There could be an alternative that `extern` in ≤2021 *won't* emit `unsafe_code`  (only compat lint) unless it contained at least one explicitly `safe`/`unsafe` item.